tls-client/HelloHttpsClient.cpp

```diff
@@ -0,0 +1,431 @@
+
+    ret = mbedtls_ssl_config_defaults( &ssl_conf, MBEDTLS_SSL_IS_CLIENT,
+                                       MBEDTLS_SSL_TRANSPORT_STREAM,
+                                       MBEDTLS_SSL_PRESET_DEFAULT );
+    if( ret != 0 )
+    {
+        logTlsError( "mbedtls_ssl_config_defaults()", ret );
+        return( ret );
+    }
+
+    mbedtls_ssl_conf_ca_chain( &ssl_conf, &cacert, NULL );
+    mbedtls_ssl_conf_rng( &ssl_conf, mbedtls_ctr_drbg_random, &ctr_drbg );
+
+    /*
+     * It is possible to disable authentication by passing
+     * MBEDTLS_SSL_VERIFY_NONE in the call to mbedtls_ssl_conf_authmode()
+     */
+    mbedtls_ssl_conf_authmode( &ssl_conf, MBEDTLS_SSL_VERIFY_REQUIRED );
+
+#if HELLO_HTTPS_CLIENT_DEBUG_LEVEL > 0
+    mbedtls_ssl_conf_verify( &ssl_conf, sslVerify, NULL );
+    mbedtls_ssl_conf_dbg( &ssl_conf, sslDebug, NULL );
+    mbedtls_debug_set_threshold( HELLO_HTTPS_CLIENT_DEBUG_LEVEL );
+#endif /* HELLO_HTTPS_CLIENT_DEBUG_LEVEL > 0 */
+
+    ret = mbedtls_ssl_setup( &ssl, &ssl_conf );
+    if( ret != 0 )
+    {
+        logTlsError( "mbedtls_ssl_setup()", ret );
+        return( ret );
+    }
+
+    ret = mbedtls_ssl_set_hostname( &ssl, server_name );
+    if( ret != 0 )
+    {
+        logTlsError( "mbedtls_ssl_set_hostname()", ret );
+        return( ret );
+    }
+
+    mbedtls_ssl_set_bio( &ssl, static_cast<void *>( socket ), sslSend,
+                         sslRecv, NULL );
+
+    return( 0 );
+}
+
+int HelloHttpsClient::sslRecv( void *ctx, unsigned char *buf, size_t len )
+{
+    TCPSocket *socket = static_cast<TCPSocket *>( ctx );
+    int ret = socket->recv( buf, len );
+
+    if( ret == NSAPI_ERROR_WOULD_BLOCK )
+    {
+        return( MBEDTLS_ERR_SSL_WANT_READ );
+    }
+    else if( ret < 0 )
+    {
+        mbedtls_printf( "Failed call to socket->recv(): %d\r\n", ret );
+    }
+
+    return( ret );
+}
+
+int HelloHttpsClient::sslSend( void *ctx, const unsigned char *buf, size_t len )
+{
+    TCPSocket *socket = static_cast<TCPSocket *>( ctx );
+    int ret = socket->send( buf, len );
+
+    if( ret == NSAPI_ERROR_WOULD_BLOCK )
+    {
+        return( MBEDTLS_ERR_SSL_WANT_WRITE );
+    }
+    else if( ret < 0 )
+    {
+        mbedtls_printf( "Failed call to socket->send(): %d\r\n", ret );
+    }
+
+    return( ret );
+}
+
+void HelloHttpsClient::sslDebug( void *ctx, int level, const char *file,
+                                 int line, const char *str )
+{
+    (void)ctx;
+
+    const char *p, *basename;
+
+    /* Extract basename from file */
+    for( p = basename = file; *p != '\0'; p++ )
+    {
+        if( *p == '/' || *p == '\\' )
+        {
+            basename = p + 1;
+        }
+    }
+
+    mbedtls_printf( "%s:%d: |%d| %s\r", basename, line, level, str );
+}
+
+int HelloHttpsClient::sslVerify( void *data, mbedtls_x509_crt *crt, int depth,
+                                 uint32_t *flags )
+{
+    (void)data;
+
+    int ret = -1;
+    char *buf = new char[GENERAL_PURPOSE_BUFFER_LENGTH];
```

Comment:
Is there a reason we cannot use ```gp_buf```` here? It is fine if we decide to use a localised ```gp_buf```` at the other place.